crates/ty_ide/src/rename.rs

```diff
@@ -0,0 +1,773 @@
+    let parsed = ruff_db::parsed::parsed_module(db, file);
+    let module = parsed.load(db);
+
+    // Get the definitions for the symbol at the offset
+    let goto_target = find_goto_target(&module, offset)?;
+
+    // Don't allow renaming of import module components
+    if matches!(
+        goto_target,
+        crate::goto::GotoTarget::ImportModuleComponent { .. }
+    ) {
+        return None;
+    }
+
+    let current_file_in_project = is_file_in_project(db, file);
+
+    if let Some(definition_targets) =
+        goto_target.get_definition_targets(file, db, None, ImportAliasResolution::PreserveAliases)
+    {
+        for target in &definition_targets {
+            let target_file = target.file();
+
+            // If definition is outside the project, refuse rename
+            if !is_file_in_project(db, target_file) {
+                return None;
+            }
+
+            // If current file is not in project and any definition is outside current file, refuse rename
+            if !current_file_in_project && target_file != file {
+                return None;
+            }
+        }
+    } else {
+        // No definition targets found. This happens for keywords, so refuse rename
+        return None;
+    }
+
+    Some(goto_target.range())
+}
+
+/// Perform a rename operation on the symbol at the given position.
+/// Returns all locations that need to be updated with the new name.
+pub fn rename(
+    db: &dyn Db,
+    file: File,
+    offset: TextSize,
+    new_name: &str,
+) -> Option<Vec<ReferenceTarget>> {
+    let parsed = ruff_db::parsed::parsed_module(db, file);
+    let module = parsed.load(db);
+
+    // Get the definitions for the symbol at the offset
+    let goto_target = find_goto_target(&module, offset)?;
+
+    // Clients shouldn't call us with an empty new name, but just in case...
+    if new_name.is_empty() {
+        return None;
+    }
+
+    // Determine if we should do a multi-file rename or single-file rename
+    // based on whether the current file is part of the project
+    let current_file_in_project = is_file_in_project(db, file);
+
+    // Choose the appropriate rename mode:
+    // - If current file is in project, do multi-file rename
+    // - If current file is not in project, limit to single-file rename
+    let rename_mode = if current_file_in_project {
+        ReferencesMode::RenameMultiFile
+    } else {
+        ReferencesMode::Rename
+    };
+
+    // Find all references that need to be renamed
+    references(db, file, &goto_target, rename_mode)
+}
+
+/// Helper function to check if a file is included in the project.
+/// Returns `false` if the file path cannot be converted to a system path.
+fn is_file_in_project(db: &dyn Db, file: File) -> bool {
+    let project = db.project();
+    let file_path = file.path(db);
+
+    file_path
+        .as_system_path()
+        .map(|path| project.is_file_included(db, path))
+        .unwrap_or(false)
```

Comment:
`is_file_included` is not the right method here. It only returns whether the file matches any `include` pattern and isn't excluded by an `exclude` pattern but it might incorrectly return `true` for a file that's git ignored. 

We should use the more expensive `db.files().contains(file)` instead. There's an issue to make `is_file_included` more accurate so that we don't have to index all files to determine if a file is part of a project or not